src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.semconv;
+
+import io.opentelemetry.api.common.AttributeKey;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.function.Function;
+
+public final class AttributeKeyTemplate<T> {
+
+  private final String prefix;
+  private final Function<String, AttributeKey<T>> keyBuilder;
+  private final ConcurrentMap<String, AttributeKey<T>> keysCache = new ConcurrentHashMap<>(1);
+
+  AttributeKeyTemplate(String prefix, Function<String, AttributeKey<T>> keyBuilder) {
+    this.prefix = prefix;
+    this.keyBuilder = keyBuilder;
+  }
+
+  static AttributeKeyTemplate<String> stringKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<String>(prefix, AttributeKey::stringKey);
+  }
+
+  static AttributeKeyTemplate<List<String>> stringArrayKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<List<String>>(prefix, AttributeKey::stringArrayKey);
+  }
+
+  static AttributeKeyTemplate<Boolean> booleanKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<Boolean>(prefix, AttributeKey::booleanKey);
+  }
+
+  static AttributeKeyTemplate<List<Boolean>> booleanArrayKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<List<Boolean>>(prefix, AttributeKey::booleanArrayKey);
+  }
+
+  static AttributeKeyTemplate<Long> longKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<Long>(prefix, AttributeKey::longKey);
+  }
+
+  static AttributeKeyTemplate<List<Long>> longArrayKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<List<Long>>(prefix, AttributeKey::longArrayKey);
+  }
+
+  static AttributeKeyTemplate<Double> doubleKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<Double>(prefix, AttributeKey::doubleKey);
+  }
+
+  static AttributeKeyTemplate<List<Double>> doubleArrayKeyTemplate(String prefix) {
+    return new AttributeKeyTemplate<List<Double>>(prefix, AttributeKey::doubleArrayKey);
+  }
+
+  private AttributeKey<T> createAttributeKey(String keyName) {
+    String key = prefix + "." + keyName.toLowerCase(Locale.ROOT);
```

Comment:
I have somewhat mixed feelings about lowercasing the key here. We did that for the HTTP headers, but there were some discussions about whether we should keep or ignore the case; and maybe, for other template attributes in the future we'll want to keep the case. I find that unlikely though -- I'm guessing we'll probably follow the precedent set by HTTP headers; and `key` and `Key` being different things is _super_ confusing.

Reply:
Yeah, I was basically following the HTTP headers example. But agree that for others we might need to keep the key as is. What do others think?

Reply:
Actually let me change that to not lowercasing it here and do the `toLowerCase()` in the instrumentation when using it for HTTP headers. WDYT?

Reply:
> Actually let me change that to not lowercasing it here and do the `toLowerCase()` in the instrumentation when using it for HTTP headers. WDYT?

👍 
Should be a bit more future-proof in case the semconv decides to use case sensitive naming somewhere (I hope they don't 😅 )